### DISCUSSION TITLE
Is it safe to manually set `abortStatusCode`?

### DISCUSSION THREAD

**Discussion by develohpanda** [vikejs/vike#1492] [Discussion]
Hiya!

The only way I can find to alter the returned HTTP Status Code, and have the Vike logs also reflect the code, is to set `pageContext.abortStatusCode` manually. I have an integration with React Router 5 (_not_ a deep integration via `onBeforeRender`), which appears to work as intended. I am adapting their documentation for server rendering, specifically for [updating the status code](https://v5.reactrouter.com/web/guides/server-rendering/404-401-or-any-other-status).

In their example, they access a context object on the `StaticRouter` like so:

```js
function Status({ code, children }) {
  return (
    <Route
      render={({ staticContext }) => {
        if (staticContext) staticContext.status = code;
        return children;
      }}
    />
  );
}
```

I have adapted this to be
```js
function Status({ code, children }) {
  const pageContext = usePageContext();
  pageContext.abortStatusCode = code;
  return children;
}
```

In my express server, I can do something like `res.status(staticContext.status || httpResponse.statusCode)`, which also works correctly however vike logs won't reflect that.

Is there a different way to override the status code in Vike-land, other than using `throw render(404)`?

**Comment by brillout** [comment]
As you probably already figued out, the usual way is to define a `_error/+Page.js` and use `throw render(404)`. Since it isn't recommended to use Vike with React Router, having first-class support for it isn't top priority at the moment.

Alternatively to what you're currently doing, a potentially clearer way would be to set a custom property `pageContext.myPropToSetStatusCode`  and then set the `404` status code at your `renderPage()` server integration accordingly.

I'm curious: why do you want to use React Router?

**Reply by develohpanda** [reply]
My only concern with using a property other than `abortStatusCode` is that Vike logs do not show the correct response code. Is there a way to control that? Do you think it may be possible with the upcoming `onLog()` hook?

> I'm curious: why do you want to use React Router?

I’m sticking with React Router because I’m migrating a project that currently uses React Router into Vike. I had descoped replacing React Router with Vike’s routing for the MVP because that began to open a can of worms (within our codebase) when I first attempted it. Sticking with React Router is just a hold-over until I can spend more dedicated time investigating it after a MVP.

I did attempt to use `throw render(404)` even with React Router, and it sort of worked, but clicking any links from the `_error/+Page.js` didn’t route as expected.

**Reply by develohpanda** [reply]
The more I think about this, I understand that overriding the status code isn’t actually necessary with Vike’s router hence why I’m in this predicament. But I do feel there is a case for providing the flexibility to modify a status code should a project choose to use a different routing methodology. I wonder if it can be another option to the return of `onRenderHtml()`. 🤔

**Reply by brillout** [reply]
> The more I think about this, I understand that overriding the status code isn’t actually necessary with Vike’s router hence why I’m in this predicament.

Exactly: Vike fully owns the status (code) of the page since Vike owns the routing.

> I’m sticking with React Router because I’m migrating a project that currently uses React Router

Makes senses and that's very much a use case we care about.

> But I do feel there is a case for providing the flexibility to modify a status code should a project choose to use a different routing methodology. I wonder if it can be another option to the return of `onRenderHtml()`. 🤔

I agree. That said, it isn't top priority right now.

FYI in case that's an option for your company, you can consider [sponsoring](https://github.com/vikejs/vike/issues/1350) and get increased priority. Regardless, let me know how it goes and definitely let me know if you run into any issues.

**Reply by develohpanda** [reply]
Just following up that I haven’t faced any obvious issues overriding `abortStatusCode` 😊

There is another case for being able to configure the status code shown in logs as well. If using an external router like react-router, for a static render you provide a staticRenderContext, which contains information about any redirects, which is then send in the response from the server.

Currently, in that case as well, Vike will log a 200 response, and not 301/302 indicating the redirect.

I have and will continue to ask about sponsoring in general. This particular one is more a quality of life thing, but would allow for improved integration with external routers 🥳 I will endeavour to contribute examples for useful integrations I’ve implemented (for example, with i18next and vite-plugin-pwa)

**Reply by brillout** [reply]
Thanks for circling back on this. I'm glad it works out.

I agree things can definitely be polished. It could be nice to create a new feature request that lists all shortcomings.

Thanks for pushing for sponsoring it makes a tremendous difference :green_heart:

> I will endeavour to contribute examples

Sounds great :eyes:
